### ISSUE TITLE
LiveFit Sequence Numbering

### ISSUE THREAD

**Issue by teddyrendahl** [NSLS-II/bluesky#684]
Building on top of the LiveFit callback for a few applications and running into an issue if my subscription doesn't start at the first event document in the Run. Currently LiveFit calls .update_fit() if the 'seq_num' key is higher than the number of constants. 

https://github.com/NSLS-II/bluesky/blob/master/bluesky/callbacks/core.py#L758

Is there a reason that we can't just check the length of the cached array?
```python
i  = len(self.ydata)
```
I'm trying to write an averaging fit callback which is how I stumbled upon this issue, but I would imagine it would happen if you subscribed the callback mid-run as well.

**Comment by tacaswell** [comment]
Checking the sequence number is less work :smiling_imp: 

We spent a fair amount of time talking about late subscriptions and came down on it is generally a bad idea  as jumping into a stream part way through means you may see documents without seeing required preceding documents (ex, events without descriptors and descriptors without starts).  That _seems_ easy enough to fix by caching and letting lat subscribers catch up, but at that point you have started to re-implement a message queue with topics (like kafka) and you should just use one of the existing message queues (and persistently subscribe the insert all documents into the queue).

**Comment by danielballan** [comment]
@teddyrendahl I'd be interested to hear more about the use case for subscribing mid-run. It's probably at least loosely related to some things I am working on now.

**Comment by teddyrendahl** [comment]
**TLDR:** I want to apply a set of filters on my event documents so that certain ones are not included in either my `LiveFit` or `LivePlot`. I have hacked through this by updating the `seq_num` attribute of each doc as it passes through my filters to account for dropped shots.

A little bit of background, our main motivation for using bluesky is for some up and coming automation projects at LCLS. We decided the easiest way to do this was to essentially create a plan that directly incorporated a `LiveFit` into an `adaptive_scan`. This means you give a `lmfit.Model` that you think will correctly fit your data, a motor scans the parameter space at first making naive steps but slowly gathering more data, at each checkpoint checking with the fit what the estimated move is to reach a target. 

In theory this all works very well! But the LCLS FEL is very unstable, even when using the builtin suspenders, if we are reading from a detector as fast as possible, we get some "bad shots" that mess up the fit. Since we are doing all of this live we don't have the benefit of just dropping these shots in offline analysis, they have to be excluded via filters on the ``LiveFit``. 

I've been thinking about how I could generalize some of these thoughts and actually make a PR that you guys could use, but I always thought we were kind of distorting your original layout by using a callback to inform the next step of a plan. I do think there could be some interesting things done with the ``count`` plan. For instance, in the current schema if your offline analysis required 15 'good events', but these events only happened 50% of the time, you would just make a count plan that had 30 reads and deal with it offline. What I see as advantageous, and what we have in our code, is that you actually just give the set of filters to the count plan. That way the plan repeatedly requests data from the detectors until you are sure you have enough data to move on to the next step in the plan.

**Comment by tacaswell** [comment]
That makes lots of sense, :+1: to that change.

**Comment by danielballan** [comment]
Agreed.

**Comment by danielballan** [comment]
Closed by #845
